### ISSUE TITLE
roaring_bitmap_run_optimize() issue

### ISSUE THREAD

**Issue by ancuop** [RoaringBitmap/CRoaring#23]
Hi Lemire,

I confuse on function  roaring_bitmap_run_optimize(). Could you please help to explain me??
I made a simple test (please see below code), then run in 2 contexts:
1.    Build and run below code normally.
2.    Comment function // roaring_bitmap_run_optimize(reg). Then build and run again

While program is looping, i measure memory usage by command "pmap  (pid_of_program)"
I found that size of "reg" reduces much however total memory gotten from pmap is not changed.

I think memory gotten from pmap should reduce after run optimize, is it correct?

```c
int main(int argc, char** argv) 
{
    int i, k, n;
    roaring_bitmap_t *reg;
    uint32_t count;

    reg = roaring_bitmap_create();
    for (i=0; i<320000000; i++) {
        roaring_bitmap_add(reg, i);
    }
    count = roaring_bitmap_portable_size_in_bytes(reg);
    printf("size before optimize: %d\n", count);
    roaring_bitmap_run_optimize(reg);  // comment this function to run second test
    count = roaring_bitmap_portable_size_in_bytes(reg);
    printf("size after optimize: %d\n", count);
    while (1);
}
```

**Comment by fsaintjacques** [comment]
Looking quickly at `convert_run_optimize` there's a lot of potential to memleaks in this function.

**Comment by lemire** [comment]
@ancuop  Thanks for your report. This could be important.

Can you provide a reproducible testing protocol, ideally using a script, with all your software... along with a validation of your protocol. For example, write a program that uses x bytes (maybe with a single malloc call) and check that you can indeed measure accurately the memory usage of the toy program. Then doing the same with roaring bitmaps.

Without a verified and trusted protocol, it is going to be hard to work on this issue.

**Comment by owenkaser** [comment]
It looks like the new change to convert_run_to_efficient_container (no
longer automatically frees a converted run container) will lead to a leak,
if a run container is converted to a bitset or array (which should not be
happening in his scenario, I think).

Is pmap the right tool for this measurement?  If the heap has been expanded
and now lots of it is unused (after run_optimize), will we see this?

**Comment by lemire** [comment]
> It looks like the new change to convert_run_to_efficient_container (no
longer automatically frees a converted run container) will lead to a leak

Though you should feel free to revert my latest changes, I fixed this particular issue (just because I hate having broken code)

https://github.com/RoaringBitmap/CRoaring/commit/004efbf45a39370926b1c0862bde3c9870fc092f

**Comment by fsaintjacques** [comment]
Valgrind is the right tool. Take the snippet provided by @ancuop, remove the `while(1)` and free the bitmap. Valgrind will complain if there's a leak.

**Comment by fsaintjacques** [comment]
I also suspect that @owenkaser is right about pmap, it's not guaranteed that heap memory will be reclaimed from RSS after a free() call (it'll be open for subsequent call).  I don't see any direct problem:

```
$ valgrind tests/issue-23
==4661== Memcheck, a memory error detector
==4661== Copyright (C) 2002-2015, and GNU GPL'd, by Julian Seward et al.
==4661== Using Valgrind-3.11.0 and LibVEX; rerun with -h for copyright info
==4661== Command: tests/issue-23
==4661==
size before optimize: 4009808
size after optimize: 6912
==4661==
==4661== HEAP SUMMARY:
==4661==     in use at exit: 0 bytes in 0 blocks
==4661==   total heap usage: 10,296 allocs, 10,296 frees, 22,209,516 bytes allocated
==4661==
==4661== All heap blocks were freed -- no leaks are possible
==4661==
==4661== For counts of detected and suppressed errors, rerun with: -v
==4661== ERROR SUMMARY: 0 errors from 0 contexts (suppressed: 0 from 0)
valgrind tests/issue-23  20.48s user 0.04s system 99% cpu 20.516 total
printerdownstair2:croaring/ (master✗) $ cat tests/issue-23.c
#include "roaring.h"

int main() {
    roaring_bitmap_t* bitmap = roaring_bitmap_create();

    for (size_t i = 0; i < 3200000; i++) {
        roaring_bitmap_add(bitmap, i);
    }

    printf("size before optimize: %zu\n",
           roaring_bitmap_portable_size_in_bytes(bitmap));

    roaring_bitmap_run_optimize(bitmap);

    printf("size after optimize: %zu\n",
           roaring_bitmap_portable_size_in_bytes(bitmap));

    roaring_bitmap_free(bitmap);

    return EXIT_SUCCESS;
}
```

**Comment by lemire** [comment]
@fsaintjacques 

Valgrind used to be our best last hope, but I think sanitize flags are superior :

http://lemire.me/blog/2016/04/20/no-more-leaks-with-sanitize-flags-in-gcc-and-clang/

**Comment by lemire** [comment]
@owenkaser  @fsaintjacques 

Right. So one should verify, using a simple program that uses only ``malloc`` and ``free`` that pmap does indeed track the memory usage. Without such a demonstration, it will be very hard to make progress.

@ancuop   : Do you think you can provide such a demonstration?

**Comment by fsaintjacques** [comment]
@lemire, made it work with `-fsanitize=address` but it founds problem elsewhere ;)

**Comment by lemire** [comment]
@fsaintjacques  Our code still has problems. I am puzzling secretly over some of them right now. ;-)

**Comment by fsaintjacques** [comment]
Should I commit the `-fsanitize` support in cmake?

**Comment by lemire** [comment]
@fsaintjacques  I am not sure what you refer to. We already use these sanitize flags (the ones I mention in my blog post). That's how I test our code.

Right now, if you turn on SANITIZE, and then hit ``make test`` it will test things out.

Of course, we could use *more* unit tests.

**Comment by fsaintjacques** [comment]
I didn't see it! I would probably change it to be enabled by default for Debug release mode and applied by default on tests.

**Comment by lemire** [comment]
@fsaintjacques 

The README.md file instructs us to enable SANITIZE for debug. It could indeed be a default as long as it can overwritten. Note that for release, it should be optional.

**Comment by lemire** [comment]
@fsaintjacques 

Note that if we ever use travis, something we should do at some point, we cannot be sure that they will support a sufficiently recent version of the compiler.

Eventually though, I think that all C and C++ projects should be tested with these flags, not just CRoaring. This will surely catch many hidden bugs.

**Comment by fsaintjacques** [comment]
You can work around travis' lack of compiler supports:

https://travis-ci.org/fsaintjacques/libtwiddle

**Comment by lemire** [comment]
@fsaintjacques 

As you demonstrate very well, you can using ``sudo``.

**Comment by ancuop** [comment]
@lemire , @fsaintjacques, @owenkaser 
Thank you very much for your comment.

Firstly, the purpose of my test is to confirm whether run_optimize optimize memory of program or not.
My process is to run program without calling free function until the program loops in while(), then I measure memory which program is using 

From your discussion, I found that I misunderstood on pmap, so my test result is not correct.
I change to use valgrind tool to test again. 
My process is to run my simple code with valgrind (comment while(1)). When program ends, i check memory leak from valgrind. I think memory leak is the memory program allocated while it was running. 

The result is
`============= Run WITH optimize
==5564== HEAP SUMMARY:
==5564==     in use at exit: 162,138 bytes in 9,771 blocks
==5564==   total heap usage: 102,590 allocs, 92,819 frees, 221,871,540 bytes allocated
==5564== 
==5564== LEAK SUMMARY:
==5564==    definitely lost: 0 bytes in 0 blocks
==5564==    indirectly lost: 0 bytes in 0 blocks
==5564==      possibly lost: 0 bytes in 0 blocks
==5564==    still reachable: 162,138 bytes in 9,771 blocks
==5564==         suppressed: 0 bytes in 0 blocks


============= Run WITHOUT optimize
==5417== HEAP SUMMARY:
==5417==     in use at exit: 40,144,142 bytes in 9,771 blocks
==5417==   total heap usage: 92,824 allocs, 83,053 frees, 221,773,880 bytes allocated
==5417== 
==5417== LEAK SUMMARY:
==5417==    definitely lost: 0 bytes in 0 blocks
==5417==    indirectly lost: 0 bytes in 0 blocks
==5417==      possibly lost: 0 bytes in 0 blocks
==5417==    still reachable: 40,144,142 bytes in 9,771 blocks
==5417==         suppressed: 0 bytes in 0 blocks`

In conclusion, run_optimize saves memory of my program.
Once again, thank you very much!

**Comment by lemire** [comment]
Thank you.  It is clever to use ``valgrind`` for this purpose. I am closing this.
